fix: check if application has prune only changes and automated prune is disabled when doing progressive sync (Fixes #21528)

Homulvas opened the pull request:
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

Fixes #21528
This essentially does the same logic check that application controller does when autosync is enabled https://github.com/argoproj/argo-cd/pull/3253/files to check that progressive sync is actually needed

crenshaw-dev commented on applicationset/controllers/applicationset_controller.go:
```diff
@@ -1095,7 +1095,20 @@ func (r *ApplicationSetReconciler) updateApplicationSetApplicationStatus(ctx con
+			pruneEnabled := app.Spec.SyncPolicy != nil && app.Spec.SyncPolicy.Automated != nil && app.Spec.SyncPolicy.Automated.Prune
+			if !pruneEnabled {
+				requirePruneOnly := true
+				for _, r := range app.Status.Resources {
+					if r.Status != argov1alpha1.SyncStatusCodeSynced && !r.RequiresPruning {
+						requirePruneOnly = false
+						break
+					}
+				}
```
@rumstead this would introduce another dependency on the in-CR resource status. How are you feeling about the Redis solution?

rumstead replied:
thankfully progressive syncs only look at the top level application sync and health status. so we won't need any changes to the existing logic. the top level statuses are set by the application controller and aren't stored in redis.

rumstead replied:
this change should also be fine, the `app.status.resources` are still populated when storing changes in redis, just the `app.status.resources[].health` would be missing. 

eg
```
status:
  health:
    status: Healthy
  resources:
  - group: apps
    health:
      status: Healthy
    kind: Deployment
    name: demo-app-kust
    namespace: cluster-generator
    status: OutOfSync
    version: v1
```
vs
```
status:
  health:
    status: Healthy
  resourceHealthSource: appTree
  resources:
  - group: apps
    kind: Deployment
    name: demo-app-kust
    namespace: cluster-generator
    status: OutOfSync
    version: v1
```

Homulvas commented:
I've rebased and changed the logic to not even check the resources if app is not out of sync.
Let me know if anything is missing before this can be merged.

andrii-korotkov-verkada commented on applicationset/controllers/applicationset_controller.go:
```diff
@@ -1068,7 +1068,20 @@ func (r *ApplicationSetReconciler) updateApplicationSetApplicationStatus(ctx con
+						requirePruneOnly = false
+						break
+					}
+				}
```
This may be too slow, resulting in O(N * M) time complexity from the number of apps and number of resources. I think we have to do incremental updates of the target value to get linear complexity.

Homulvas replied:
I do agree that it would make things much easier if the app status had a distinction between `OutOfSync` and `OutOfSyncPruneOnly`.
At the same time if Application Controller can live with this loop https://github.com/argoproj/argo-cd/blob/4a1d0f3af5cb1f093b3b17a82093f48f36959b99/controller/appcontroller.go#L2060C1-L2072C3 then I don't see any point keeping this out from Appset Controller when it only affects Progressive Syncs.

andrii-korotkov-verkada replied:
We should probably update application status to have this field and re-compute it.

agaudreault commented:
I think the infinite loop will be fixed by https://github.com/argoproj/argo-cd/pull/23187. So I am not sure this PR will be necessary. The infinite loop happens for any use cases where the application would stay OutOfSync. Only solving it for unpruned resources is only a partial fix and does not resolve the root cause.
